src/main/jbake/content/blog/2016/07/05/scaling-hawkular-alerting.adoc

```diff
@@ -0,0 +1,108 @@
+Hawkular Alerting is the component responsible to detect behaviours on the hawkular system by defining rules.
+These rules can generate events or alerts that can also be used in additional rules to detect complex scenarios and to respond with customized actions.
+
+This highlevel process is stateful by nature in the sense that rules might depend on the past and future data received.
+
+== Scaling: fault tolerance vs distributed computation
+
+A first (a non trivial) question is what means 'scaling' in our problem domain.
+For example, 'scaling' in some traditional databases scenarios could mean to add more resources in a big backend machine, on the contrary scaling in a network means to have more nodes to distribute load and increase the computational power.
+Also another preliminar point is that we add more nodes to increase the fault tolerance of our system, meaning that we add more nodes to replicate our load, but not necessary to increase the cappacity of the system.
+For example, some networks should be able to support more than one concurrent failure of the system.
+
+For our specific Hawkular Alerting domain scaling means to have more computational cappacity but also offering a minimal fault tolerance cappacity.
+
+== PartitionManager interface
+
+Hawkular Alerting implements its main logic in the engine service, the engine is responsible to manage the rules definitions and match the data and events with the rules updating its internal state.
+The engine is the main actor to use the scaling services and in this context the scaling/clustering services have been designed from an abstract interface API called PartitionManager.
+Main benefit of this abstraction layer is that we can improve in the future the implementation of the distributed scenarios without bring technical debt into the main logic of Hawkular Alerting.
+
+The main strategy of the design is to partition the rules across all the nodes of the topology following a consistent hashing algorithm.
+This means that when topology changes the PartitionManager minimizes the tasks needed to reconfigure the state.
+
+So, in a high level, the PartitionManager interface is the responsible of the following clustering services:
+
+- Notify when a new rule (called Trigger in Hawkular Alerting context) has been added, modified or removed in order to update the engine state.
+- Detect changes on the topology to reconfigure the partition of the triggers in a transparent way when a node is added/removed from the cluster.
+- Notify when a new data/event has been received in order to route it to the engine node responsible of its processing.
+
+== Some implementation details
+
+The goal of a scaling design is efficiency, it should distribute the rules in a way that minimize the changes and cost of re-distribution when cluster changes.
+In our PartitionManagerImpl we have used the of hashing functions provided for the *guava* library to implement a shared topology as it is shown in the next example.
```

Comment:
"we have used the hashing..." (removed 'of')
"provided by the *guava*" - (replaced 'for' with 'by')